File: astar_breakdown.py
import time
import random
import torch

# # define the size of the grid
# size = 5

# # grid layout building parts
# vertical = "|"
# horizontal = "_"
# intersection = "+"


# # creating the grid
# for i in range(size):
#     # print horizaontal lines
#     print(intersection + horizontal*size + intersection)

#     # print vertical lines
#     print((vertical + " "*size)* size + vertical)

# # print the last horizontal line
# print(intersection + horizontal*size + intersection)


# end of drawing a grid
class Grid:
    def __init__(self, size):
        self.size = size
        self.grid = [["  " for _ in range(size)] for _ in range(size)]
        # position of the x
        self.position = [size // 2, size // 2]
        # for the target? why should it be in the init function?
        self.place_target()
        # place X after the target
        self.grid[self.position[0]][self.position[1]] = " X "

        # intialize the cost to zero here
        self.cost = 0



    def print_grid(self):
        for row in self.grid:
            print("|" + "|".join(row) + " | ")
            # seperator line
            print("--" * (self.size * 2 + 1))
        # wait for 1 second to print the next grid
        time.sleep(1)
        print()


    def move_up(self):
        # what does this mean?
        # going up is possible -> there is a top row available
        if self.position[0] > 0:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[0] -= 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1

    # diagonal move function towards the top right quadrant
    def move_up_right(self):
        if self.position[0] > 0 and self.position[1] < self.size - 1:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[0] -= 1
            self.position[1] += 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1



    def move_left(self):
        # able to go a left
        # draw a 5 x 5 grid on paper and see
        # there is a 1 column on the left for traversal
        if self.position[1] > 0:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[1] -= 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1


    def move_up_left(self):
        if self.position[0] > 0 and self.position[1] < self.size - 1:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[0] -= 1
            self.position[1] -= 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1


    def move_down(self):
        # able to go down by 1 unit
        # draw a 5 by 5 grid on paper and see
        # there is 1 row at the bottom for the X to go
        if self.position[0] < self.size - 1:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[0] += 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1


    def move_down_right(self):
        if self.position[0] > 0 and self.position[1] < self.size - 1:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[0] += 1
            self.position[1] -= 1


    def move_right(self):
        if self.position[1] < self.size - 1:
            self.grid[self.position[0]][self.position[1]] = "  "
            self.position[1] += 1
            self.grid[self.position[0]][self.position[1]] = " X "
            self.cost += 1

    def move_down_left(self):
        pass

    def place_target(self):
        while True:
            # generate random row and column indices
            # make sure to import random
            target_row = random.randint(0, self.size - 1)
            target_col = random.randint(0, self.size - 1)

            # if the selected cell is not currently occupied by "X", place the target
            if self.grid[target_row][target_col] == "  ":
                self.grid[target_row][target_col] = " T "
                break


    def print_cost(self):
        print(f"Total cost: {self.cost}")


    def follow_target(self):
        # scan the position of the target here
        target_position = [(i, j) for i in range(self.size) for j in range(self.size) if self.grid[i][j] == " T "][0]

        # move horizontally towards the target
        while self.position[1] != target_position[1]:
            if self.position[1] < target_position[1]:
                self.move_right()
                # self.print_grid()
            else:
                self.move_left()
                # self.print_grid()
            self.print_grid()


        # move vertically towards the target
        while self.position[0] != target_position[0]:
            if self.position[0] < target_position[0]:
                self.move_down()
            else:
                self.move_up()
            self.print_grid()



# # Initializing the grid
grid = Grid(5)

# # printing the grid before the grud
# grid.print_grid()

# # perform the movement operation
# grid.move_up()

# # print the grid after
# grid.print_grid()

# print the left movement operation
# grid.move_left()

# print the grid after
# grid.print_grid()

# print the down movement operation
# grid.move_down()

# print the grid
# grid.print_grid()

# print the right movement operation
# grid.move_right()

# print the grid
# grid.print_grid()


# put a target on the grid
# grid.place_target()


# print the grid
grid.print_grid()

# move the target
# grid.follow_target()


# grid.print_cost()


# print the diagonal movement functions out!
grid.move_up_left()

grid.print_grid()


# adding the cost part to the algorithm? why is it important though?

# basically cost is a measure of the effort thats required for the X

# 1. path optimization -> in more complex scenarios when there are multiple routes to the target?
# this becomes important. maybe increase the grid size? and check this point further down the road

# 2. reward system
# this basically comes under a reinforcement algorithm the cost function can serve as a reward system
# that encourages the AI to find an optimal path

# 3. energy constraints
# if your simulating a real world situation, the x might represent a
# robot that has a finite amount of energy. in such cases minimizing the cost
# would be critical -> very complex scenario!


# remember our basic cost function also resembles the steps taken here, so its a pretty basic system here

File: a_star_breakdown.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <unistd.h>

#define size 5
#define empty "  "
#define target " T "
#define pos " X "

// declare the grid and the position
char grid[size][size][3];
int position[2];

void print_grid()
{
    for (int i = 0; i < size; i++)
    {
        for (int j= 0; j < size; j++)
        {
            printf("|%3s", grid[i][j]);
        }
        printf("|\n");
        printf("---------------------\n");
    }
    // sleep for a second
    sleep(1);
    printf("\n");
}


// for reference
// origin is at the top left

//     0   1   2   3   4   <- x-coordinate (position[1])
// 0 | a | b | c | d | e |
// 1 | f | g | h | i | j |
// 2 | k | l | m | n | o |
// 3 | p | q | r | s | t |
// 4 | u | v | w | x | y |
// ^
// |
// y-coordinate (position[0])


void initialize_grid()
{
    position[0] = size/ 2;
    position[1] = size/ 2;

    // initialize all cells to empty
    for(int i =0; i < size; i++)
    {
        for(int j = 0; j < size; j++)
        {
            strcpy(grid[i][j] , empty);
        }
    }

    // set target to a random position
    int target_row = rand() % size;
    int target_col = rand() % size;

    strcpy(grid[target_row][target_col], target);

    // set x position
    strcpy(grid[position[0]][position[1]], pos);


}

void move_right()
{
    if(position[1] < size - 1)
    {
        // ensure we are not on the right edge of the grid
        // make current cell empty
        strcpy(grid[position[0]][position[1]], empty);

        // move to the right
        position[1] += 1;

        // mark new cell as pos
        strcpy(grid[position[0]][position[1]], pos);
    }
    else {
        printf("Cant move right from here\n");
    }
}


void move_left()
{
    if(position[1] > 0)
    {
        strcpy(grid[position[0]][position[1]], empty);

        position[1] -= 1;

        strcpy(grid[position[0]][position[1]], pos);
    }
    else {
        printf("canot move left from here\n");
    }
}







int main()
{
    printf("create the grid here\n");
    srand(time(NULL));
    initialize_grid();
    print_grid();

    // printf("move right\n");
    // move_right();
    // print_grid();

    printf("move left\n");
    move_left();
    print_grid();

    return 0;
}